controlplane/kubeadm/internal/controllers/status.go

```diff
@@ -135,3 +142,276 @@ func (r *KubeadmControlPlaneReconciler) updateStatus(ctx context.Context, contro
+		var message string
+		if missingReferencesMessage != "" {
+			message = fmt.Sprintf("Scaling up would be blocked %s", missingReferencesMessage)
+		}
+		v1beta2conditions.Set(kcp, metav1.Condition{
+			Type:    controlplanev1.KubeadmControlPlaneScalingUpV1Beta2Condition,
+			Status:  metav1.ConditionFalse,
+			Reason:  controlplanev1.KubeadmControlPlaneNotScalingUpV1Beta2Reason,
+			Message: message,
+		})
+		return
+	}
+
+	message := fmt.Sprintf("Scaling up from %d to %d replicas", currentReplicas, desiredReplicas)
+	if missingReferencesMessage != "" {
+		message = fmt.Sprintf("%s is blocked %s", message, missingReferencesMessage)
+	}
+	messages := []string{message}
+
+	if preflightChecks.HasDeletingMachine {
+		messages = append(messages, "waiting for Machine being deleted")
+	}
+
+	if preflightChecks.ControlPlaneComponentsNotHealthy {
+		messages = append(messages, "waiting for control plane components to be healthy")
+	}
+
+	if preflightChecks.EtcdClusterNotHealthy {
+		messages = append(messages, "waiting for etcd cluster to be healthy")
+	}
+
+	v1beta2conditions.Set(kcp, metav1.Condition{
+		Type:    controlplanev1.KubeadmControlPlaneScalingUpV1Beta2Condition,
+		Status:  metav1.ConditionTrue,
+		Reason:  controlplanev1.KubeadmControlPlaneScalingUpV1Beta2Reason,
+		Message: strings.Join(messages, "; "),
+	})
+}
+
+func setScalingDownCondition(_ context.Context, kcp *controlplanev1.KubeadmControlPlane, machines collections.Machines, preflightChecks internal.PreflightCheckResults) {
+	if kcp.Spec.Replicas == nil {
+		v1beta2conditions.Set(kcp, metav1.Condition{
+			Type:   controlplanev1.KubeadmControlPlaneScalingDownV1Beta2Condition,
+			Status: metav1.ConditionUnknown,
+			Reason: controlplanev1.KubeadmControlPlaneScalingDownWaitingForReplicasSetV1Beta2Reason,
+		})
+		return
+	}
+
+	currentReplicas := int32(len(machines))
+	desiredReplicas := ptr.Deref(kcp.Spec.Replicas, 0)
+	if !kcp.DeletionTimestamp.IsZero() {
+		desiredReplicas = 0
+	}
+
+	if currentReplicas <= desiredReplicas {
+		v1beta2conditions.Set(kcp, metav1.Condition{
+			Type:   controlplanev1.KubeadmControlPlaneScalingDownV1Beta2Condition,
+			Status: metav1.ConditionFalse,
+			Reason: controlplanev1.KubeadmControlPlaneNotScalingDownV1Beta2Reason,
+		})
+		return
+	}
+
+	messages := []string{fmt.Sprintf("Scaling down from %d to %d replicas", currentReplicas, desiredReplicas)}
+	if preflightChecks.HasDeletingMachine {
+		messages = append(messages, "waiting for Machine being deleted")
+	}
+
+	machinesStaleDeleting := []string{}
+	for _, machine := range machines {
+		if !machine.GetDeletionTimestamp().IsZero() && time.Since(machine.GetDeletionTimestamp().Time) > time.Minute*30 {
+			machinesStaleDeleting = append(machinesStaleDeleting, machine.GetName())
+		}
+	}
+
+	if len(machinesStaleDeleting) == 1 {
+		messages = append(messages, fmt.Sprintf("Machine %s is in deletion since more than 30m", machinesStaleDeleting[0]))
+	} else if len(machinesStaleDeleting) > 1 {
```

Comment:
No strong objection, but I have a todo task about improving consistency across aggregation messages, so most likely this will change to use a shared utility soon

Reply:
Fine either way from my side